### DISCUSSION TITLE
User Management

### DISCUSSION THREAD

**Discussion by DARB-CCM-S-20** [cloudnative-pg/cloudnative-pg#1179] [Ideas]
I would like to see the ability to natively create different users and roles for databases, maybe even providing some default roles similar to what the Zolando operator provides. 

I can see that this is in the roadmap but there is no details as to the scope, timeline or priority.

Could you provide an update in this regard? Currently this is the only thing stopping us migrating from the zolando operator to this operator

**Comment by sxd** [comment]
@DARB-CCM-S-20 it's something that we wanted to work for a while now, if you have an idea of how this should be implemented, like adding another CRD like `Users.postgresql.cnpg.io` or it's something that we should add inside the Cluster? I don't have a strong opinion on this but it's something that we mail talk in the next community meeting and you can give us your ideas on this? what do you think @gbartolini @leonardoce @mnencia

**Comment by DARB-CCM-S-20** [comment]
@sxd I think Users and their roles are something that is tied to the database so should really be database specific and instead implemented as a property of the `bootstrap` spec. However, this adds another limitation where is seems you only really have the options to bootstrap the cluster with a single `database`, where in the PostgreSQL hierarchy you should be able to have multiple databases within each cluster which itself can have multiple schemas. The only way I can see at the moment to orchestrate this is by providing your own custom SQL to the `postImportApplicationSQL`, which I think is far from ideal and would be much nicer to have a declarative approach. 

Hope that makes sense, glad to hear its on the radar.

**Comment by adminnz** [comment]
I think it would be more flexible if users could be managed via CRD and linked to a cluster and optionally a database in the cluster.
That way you can still add users via SQL, but also just setup a CRD for a new app that is being deployed that should have access to the cluster/db.

**Comment by aran** [comment]
I ended up here because one cluster per service gets expensive if there are multiple small services. I am looking for a way to map services to users to allow more than one service to use the same cluster. In this case, users/roles are a means to an end. It would be nice to consider first-class support for Postgres Certificate Authentication or other authentication schemes, because I don't *really* want to manage roles, passwords and secrets. What I really want is declaratively set up cert-manager or spiffe or something like that to handle authentication and secret rotation for supported services.

**Comment by davinkevin** [comment]
+1 for this feature, with ability to define Users (and maybe more, like grant or Role) from specific CRD type.

I've used that for years in production with ConfigConnector, and it's a pleasure to administrate this as a cluster admin: https://cloud.google.com/config-connector/docs/reference/resource-docs/sql/sqluser

Simple and powerful… and because it's separated from the database, the original object is not mutated in any way, it's used as a `ref`

**Comment by sxd** [comment]
This has already been implemented closing the discussion
